### ISSUE TITLE
What is the reason that less dataset is needed to train on your implementation?

### ISSUE THREAD

**Issue by begeekmyfriend** [Rayhane-mamah/Tacotron-2#35]
I have tried both your and Keith Ito's implementation but only 10h+ dataset is needed to train into convergence on your version while there should be 20h+ dataset on Ito's one. I know that your version is deprived from Ito's one. So I wonder which module you changed that results in the effect?

**Comment by Rayhane-mamah** [comment]
Hello @begeekmyfriend, that's a nice question actually! Must be one of life's mysteries :)

Actually, despite using the same cleaners and project structure (I love how he arranges his code), there are many differences in our models and in our preprocessing.

As a consequence I am not really sure about the reason, but it is important to note that we use much more regularization than in keith ito's work (L2, dropout, zoneout, etc.) Which would be a hint if keith ito's work diverges due to overfitting on small datasets.

Please also note that we do not train using the same loss functions and that our target distributions are not the same, which also creates a whole lot of differences between our works.

So, the short answer is I have no clear idea about the reason, but I would love to know if someone could explain it to us!

**Comment by begeekmyfriend** [comment]
I think I have found one of the factors that differs the time to achieve convergence. I modified the `_process_utterence` function in `datasets/preprocessor.py` like that:
```python
def _process_utterance(mel_dir, linear_dir, wav_dir, index, wav_path, text):
        ...
        #M-AILABS extra silence specific
        # if hparams.trim_silence:
        #       wav = audio.trim_silence(wav)

        #Mu-law quantize
        if is_mulaw_quantize(hparams.input_type):
                #[0, quantize_channels]
                ...
        elif is_mulaw(hparams.input_type):
                #[-1, 1]
                ...
        else:
                #[-1, 1]
                ...

        # Compute the mel scale spectrogram from the wav
        mel_spectrogram = audio.melspectrogram(wav).astype(np.float32)
        mel_frames = mel_spectrogram.shape[1]

        #Compute the linear scale spectrogram from the wav
        linear_spectrogram = audio.linearspectrogram(wav).astype(np.float32)
        linear_frames = linear_spectrogram.shape[1]

        #sanity check
        assert linear_frames == mel_frames

        time_steps = mel_frames * audio.get_hop_size()
        assert time_steps >= mel_frames * audio.get_hop_size()
        assert time_steps % audio.get_hop_size() == 0

        # Write the spectrogram and audio to disk
        audio_filename = 'speech-audio-{:05d}.npy'.format(index)
        mel_filename = 'speech-mel-{:05d}.npy'.format(index)
        linear_filename = 'speech-linear-{:05d}.npy'.format(index)
        ...
```
As you can see I have commented out `pad_lr` method for the training audio and I will explain why I did it. And then I conducted three experiments by contrast.
- Set `hp.input_type` as `raw` and kept `pad_lr` working;
- Set `hp.input_type` as `raw` and commented `pad_lr` out;
- Set `hp.input_type` as `mulaw-quantize` and commented `pad_lr` out;
On the first try it failed to convergence. And on the second it began to convergence at 4K steps. On the third time it achieved convergence at only 2K steps.
![step-1000-align](https://user-images.githubusercontent.com/6031938/39952340-e3901b06-55c6-11e8-972c-9f49171d7121.png)
![step-2000-align](https://user-images.githubusercontent.com/6031938/39952342-e53bb776-55c6-11e8-91c5-2c0ebfaa5884.png)
In my opinion the trim of silence really helps alignments. As we can see in Mu-law quantized input mode, we quantized the audio into 256 channels and then called `audio.start_and_end_indices` method to trim the silence in the header and the tailer accurately to unit of sample point. So it reached convergence fastest. In raw input mode we do not trim silence but the original audio does not contain much silence at both ends so it still can get to convergence. But when we add `pad_lr` method and switch to raw input mode, it actually impair the purity of the training audio. So it can never get to convergence due to this reason.
In fact, `pad_lr` does not help training. We can just set timestep as `mel_frames * hop_size` at will here since timestep is useless in the subsequent training. I think we should comment out those useless code.

**Comment by butterl** [comment]
@begeekmyfriend  Tried with thchs30 datasets and with raw mode didn't get convergence until 12400 steps(with pad_lr) and loss went down fast to 0.2x even with step 1K

outputs_per_step=3 and batch_size=16 due to small GPU memory，seems bigger outputs_per_step and bigger batch_size also help get convergence in my try to deal with OOM .Thanks for sharing your information，will try comment out pad_lr and feed back.

Also see you comments out trim_silence function in the demo code？

**Comment by PetrochukM** [comment]
@begeekmyfriend The ``pad_lr`` method is run on top of the ``wav`` not the ``mel-spectrogram``; therefore, it would in no way affect the Tacotron-2 spectrogram convergence.

**Comment by begeekmyfriend** [comment]
@PetrochukM `trim_silence` does the same thing with `audio.start_and_end_indices` but the latter trim accurately to sample points. Moreover I have tried on `THCHS30` and it got convergence (at about 4K steps with r=3 batch_size=32) but not under these modification. https://github.com/Rayhane-mamah/Tacotron-2/issues/39 shows reduction of batch size would result in failure to convergence.

@PetrochukM I do not think so. The `pad_lr` does affect the input of melspectrogram since the input is the wav. I have conducted several experiments to confirm that with my 10h dataset. So can you.

**Comment by PetrochukM** [comment]
@begeekmyfriend I believe that ``trim_silence`` affects the convergence. Earlier you mentioned you commented out ``pad_lr`` not ``trim_silence``. 

Show me in the code where ``pad_lr`` affects the ``mel_spectrogram`` please because it looks like it does not after reading the code.

**Comment by begeekmyfriend** [comment]
@PetrochukM you can see the input of [audio.melspectrogram](https://github.com/Rayhane-mamah/Tacotron-2/blob/master/datasets/preprocessor.py#L104) is `wav[start : end]`. And you can print the start and the end to see the changes.

**Comment by PetrochukM** [comment]
@begeekmyfriend Literally, ``l, r = audio.pad_lr(wav, hparams.fft_size, audio.get_hop_size())`` ``pad_lr`` is run after that line.

Are you focused on ``pad_lr`` or ``trim_silence``?

Previously, you said:
> As you can see I have commented out pad_lr method for the training audio and I will explain why I did it. And then I conducted three experiments by contrast.

I assume you meant you commented out ``trim_silence``.

**Comment by begeekmyfriend** [comment]
@PetrochukM Yes, the `pad_lr` cancels the good effect of `audio.start_and_end_indices`.

**Comment by PetrochukM** [comment]
``pad_lr`` is on line 115 while ``mel_spectrogram`` is on line 104, they do not affect each other.

I give up.

**Comment by begeekmyfriend** [comment]
Well maybe there is something wrong... I am nearly insane this morning. Let me cool down and keep observation...

**Comment by begeekmyfriend** [comment]
Hey I find a bug in `audio.trim_silence` where there is `librosa.effects.trim` with default arguments of `frame_length=2048` and `hop_length=512` while our hyper parameters with `fft_size=1024` and `hop_size=256`. I would fix into the right arguments and see whether it would get convergence within 2K steps under raw input mode.

**Comment by begeekmyfriend** [comment]
![step-2000-align](https://user-images.githubusercontent.com/6031938/39954394-4ae6bc5e-55f1-11e8-853e-12fd7a6f1c82.png)
Succeeded to get convergence at 2K steps in raw input mode with `librosa.effects.trim` fixed. @Rayhane-mamah I will shoot a PR for you.

**Comment by a3626a** [comment]
If you think `pad_lr` cancels out good effects of `trim`, I suggest you to try to pass `center=false` to `librosa.stft`. With this parameter, `pad_lr` is not required.

https://librosa.github.io/librosa/generated/librosa.core.stft.html

**Comment by begeekmyfriend** [comment]
@a3626a I was wrong. Actually `pad_lr` does nothing for audio pre-processing and training since the time step in metadata is useless for the feeder, It would still get to convergence if you set the calculated time step as zero.

**Comment by a3626a** [comment]
Right. `pad_lr` does not affect spectrograms.

**Comment by begeekmyfriend** [comment]
@Rayhane-mamah I have conducted series of ablation studies to find the reasons and here is the modification based on Keith Ito's version https://github.com/keithito/tacotron/issues/170. It confirms to me that the architecture of decoder and the normalization of mel-spectrograms is the key components which determinate whether we can use less dataset to get to convergence or not. I am still use L1 loss and `LSTMBLockCell` for minimal modification in this study. In your guess, L2, dropout, zoneout and cumulative attention states are not included but it may improve or speed up convergence. For your information.

**Comment by Rayhane-mamah** [comment]
@begeekmyfriend, Great observations!! I actually confirm your results.

I am not really sure about L2 regularization, I know it reduces overfit for sure, but can't confirm if it helps convergence. On the other hand, cumulative attention is essential for convergence, without it the model gets stuck from time to time. It's predictable considering that weights cumulation is typically adding more information to the network that without it, which makes predictions easier.

**Comment by begeekmyfriend** [comment]
@Rayhane-mamah Yes, you are right. The cumulative attention states is essential and I have verified it under `r==2` https://github.com/keithito/tacotron/issues/170. What is more, the bias in `_location_sensitive_score` is also dispensable for convergence and good alignment.
You seems did everything right in the code. Congratulations!!

**Comment by drasted** [comment]
hi @begeekmyfriend  ，小白一枚，最近刚好在找tacotron的源代码资料，看到了你的帖子 ，厉害啊，用THCHS30的数据集训练出不错的效果；想问下，对于初学者，想先出个demo看看效果，是直接用您https://github.com/begeekmyfriend/tacotron/commits/tacotron2 这个链接下的代码 配合数据集 就可以大致有一个雏形了吗？ 差不多训练多长时间能等到有输出的啊（用的云端GPU😂）

**Comment by begeekmyfriend** [comment]
@drasted My fork is only for ablation study and the results of synthesis has not been tested yet. You had better not use it.

**Comment by v-yunbin** [comment]
@begeekmyfriend why mine sill not converge on 6000 steps with 9.6hr thchs30 data? but your 10hr data has converge on 2000 steps?

**Comment by v-yunbin** [comment]
@begeekmyfriend why mine sill not converge on 6000 steps with 9.6hr thchs30 data? but your 10hr data has converge on 2000 steps?

**Comment by Rayhane-mamah** [comment]
@begeekmyfriend thanks for all the contributions, your help is very valuable!

@v-yunbin it's okey if you don't get alignments right away, I usually get them around 16k :)
